Feature: NetCDF support to the tifviewer

HarshShinde0 opened the pull request:
## Changes Made

- Added NetCDF file format support to the tifviewer
- Added sample NetCDF climate data file (CMIP6 dataset) https://dap.ceda.ac.uk/badc/cmip6/data/CMIP6/ScenarioMIP/KIOST/KIOST-ESM/ssp585/r1i1p1f1/Amon/tas/gr1/v20191106/tas_Amon_KIOST-ESM_ssp585_r1i1p1f1_gr1_201501-210012.nc
- Enhanced the viewer to handle temporal data with navigation controls
- Added support for climate data visualization with appropriate colormaps


Dataset link - https://dap.ceda.ac.uk/badc/cmip6/data/CMIP6/ScenarioMIP/KIOST/KIOST-ESM/ssp585/r1i1p1f1/Amon/tas/gr1/v20191106/tas_Amon_KIOST-ESM_ssp585_r1i1p1f1_gr1_201501-210012.nc

```bash
viewtif examples/sample_data/tas_Amon_KIOST-ESM_ssp585_r1i1p1f1_gr1_201501-210012.nc
```

## Demonstrating the NetCDF support:

https://github.com/user-attachments/assets/342cd7de-aaa7-4a6e-a51f-6bc611d5fad3


## Technical Details

- Added support for reading NetCDF (.nc) files using xarray
- Implemented time dimension navigation (PageUp/PageDown keys and slider)
- Added specialized colormaps for climate data (RdBu_r, viridis, magma)
- Improved geographic visualization with coastlines when cartopy is available
- Sample CMIP6 dataset: KIOST-ESM model SSP5-8.5 scenario (2015-2100)
- Enhanced time series support with play/pause animation functionality
- Added time slider for easy navigation through temporal data
- Implemented date selector dropdown for precise timestep selection
- Added automatic timestep progression with configurable animation speed

This PR extends the viewer's capabilities beyond static TIF & HDF5 files to include temporal NetCDF data commonly used in climate research and geospatial analysis.

nkeikon commented:
Thanks for this! I was actually thinking about adding NetCDF support too. This looks fantastic! 

To keep viewtif lightweight, would you be open to making the NetCDF and related dependencies optional (similar to how shapefile and HDF support are handled)? That way, users can install them only if they need .nc support, for example with pip install viewtif[netcdf].

Also, could you make sure that imports like xarray and cartopy are only loaded when a NetCDF file is actually opened, since they’re quite heavy dependencies?

The sample dataset link is great. Maybe we can just reference it in the README instead of adding the full file to the repo.

HarshShinde0 commented:
Hi Keiko @nkeikon,
I've completed the changes we discussed:
- NetCDF support is now optional via viewtif[netcdf]
- xarray, cartopy, and other heavy dependencies are only imported when needed
- Updated pyproject.toml with the [netcdf] extra

Let me know if you have any feedback or suggestions!

nkeikon commented:
This is excellent! It works perfectly with the sample dataset.

There seem to be a few issues (which are somewhat expected for this data format) when using different samples. My guess is that this is mostly related to the lat/lon order or the time dimension (or lack of it in some cases). I tested with our dataset (https://data.nas.nasa.gov/nex-dcp30-cmip6/data.php), and it showed this kind of behavior. Would you mind taking a look? Once it works with a few more .nc files, I think it’s ready!

HarshShinde0 commented:
Hi @nkeikon,

Thanks for the feedback!
Regarding the issues you mentioned, I’ve tested the implementation with a few additional NetCDF files and it seems to be working well so far. To better reproduce the specific issues you’re seeing, could you please share the exact dataset you used?
I tried with the dataset from this link:
https://data.nas.nasa.gov/nex-dcp30-cmip6/monthly/ACCESS-CM2/ssp370/r1i1p1f1/tasmax/
Specifically:
https://data.nas.nasa.gov/nex-dcp30-cmip6/monthly/ACCESS-CM2/ssp370/r1i1p1f1/tasmax/tasmax_mon_ACCESS-CM2_ssp370_r1i1p1f1_gn_2021.nc

<img width="1919" height="1017" alt="image" src="https://github.com/user-attachments/assets/f9553f1b-149e-4bd9-a4f0-11b1edd82feb" />

I’ve also added some fixes to improve compatibility with various NetCDF formats:
- Calculated and downsampled the lat/lon arrays to match the data shape exactly
- Used indexing='xy' for proper (lat, lon) array alignment
- Set map extent based on downsampled coordinate bounds to display only the actual data region
- Converted longitudes from 0–360 to -180/+180 for Cartopy compatibility

Also i think sometimes it might be a Cartopy boundary issue in some cases 

Let me know which file caused the behavior you saw, and I’ll make sure to test with it.
Please share it with me here  - harshinde.hks@gmail.com
Thanks again!

HarshShinde0 commented:
Hey @nkeikon, could you please reopen this PR#4 so i can push new changes here.
Thanks!
